### PR TITLE
MAINT: Fix for future change

### PR DESCRIPTION
**Pull request by bashtage**

### CODE REVIEW

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
Some ideas for improvement.  Thanks for looking into this

File: tests/test_frame.py [hunk 1]
```diff
@@ -933,12 +934,21 @@ def test_types_describe() -> None:
         }
     )
     df.describe()
-    with pytest.warns(FutureWarning, match="Treating datetime data as categorical"):
+    if PD_LTE_15:
+        with pytest.warns(FutureWarning, match="Treating datetime data as categorical"):
+            df.describe(percentiles=[0.5], include="all")
+    else:
         df.describe(percentiles=[0.5], include="all")
```
[the remark this answers is not in the document]
Comment (bashtage) [reply tests/test_frame.py]: I refactored it into a more sophisticated CM that can handle version ranges.

File: tests/__init__.py [hunk 2]
```diff
@@ -7,10 +7,14 @@
...
 )
 
+from packaging.version import parse
+import pandas as pd
+
 from pandas._typing import T
 
 TYPE_CHECKING_INVALID_USAGE: Final = TYPE_CHECKING
 WINDOWS = os.name == "nt" or "cygwin" in platform.system().lower()
+PD_LTE_15 = parse(pd.__version__) < parse("1.5.999")
```
[the remark this answers is not in the document]
Comment (bashtage) [reply tests/__init__.py:22]: I left packaging in which is required for us.  packaging.version.parse is the universal way to do this AFAICT.

File: tests/__init__.py [hunk 3]
```diff
@@ -28,3 +37,59 @@ def check(actual: T, klass: type, dtype: type | None = None, attr: str = "left")
...
+
+
+def pytest_warns_bounded(
+    warning: type[Warning],
+    match: str,
+    lower: str | None = None,
+    upper: str | None = None,
+) -> AbstractContextManager:
+    """
+    Version conditional pytet.warns context manager
```
Comment (Dr-Irv) [inline tests/__init__.py:49]: `pytest` not `pytet`

File: tests/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply tests/__init__.py:22]: Yes, but we use `pandas.util.version.Version` throughout the pandas code base (including in the pandas tests), so I'd rather follow the pandas convention.  Apparently that was adapted from `packaging` according to the source code.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
Small typo in a comment, and would prefer to use pandas `Version` class

File: tests/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bashtage) [reply tests/__init__.py]: Switched.

File: tests/__init__.py (hunk 3 above)
Comment (bashtage) [reply tests/__init__.py]: Fixed.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
Looks like we need to add a stub for `pandas.util.version`

Sorry to create more work.  Hopefully `stubgen` can get you something useful.

**Comment by Dr-Irv** [comment]
@bashtage Not sure how you want to handle this.  Looks like latest pandas nightly has an issue with matplotlib.

**Comment by bashtage** [comment]
> @bashtage Not sure how you want to handle this. Looks like latest pandas nightly has an issue with matplotlib.

I think this one is ready and it still fixes the one issue.  The MPL seems like it will be fixed upstream.

**Comment by Dr-Irv** [comment]
> > @bashtage Not sure how you want to handle this. Looks like latest pandas nightly has an issue with matplotlib.
> 
> I think this one is ready and it still fixes the one issue. The MPL seems like it will be fixed upstream.

Issue here is that the `pytest` is failing, and the typing check on nightly never happens, so maybe change the order of the tests on nightly to do the typing tests first, and then do `pytest` tests?

**Comment by bashtage** [comment]
It doesn't run typing against nightly at all -- I think this makes sense since typing is internal to pandas-stubs.  It only runs pytest to try and find issues where pytest will fail.

**Comment by Dr-Irv** [comment]
> It doesn't run typing against nightly at all -- I think this makes sense since typing is internal to pandas-stubs. It only runs pytest to try and find issues where pytest will fail.

OK.  I remembered the failure we had with `Version`, but that was in the regular checks.

**Review by Dr-Irv — APPROVED** [review]
thanks @bashtage

**Pull request merged**